firmware/src/rdm_responder.c

```diff
@@ -149,411 +120,239 @@ static void SendDUBResponseIfRequired(const uint8_t *param_data,
-                              unsigned int param_data_length) {
-  if (param_data_length != 0) {
-    SendNack(header, NR_FORMAT_ERROR);
-    return;
+int RDMResponder_SetUnMute(const RDMHeader *header) {
+  if (header->param_data_length) {
+    return RDM_RESPONDER_NO_RESPONSE;
   }
-
-  RespondIfRequired(header, ACK, GET_COMMAND_RESPONSE,
-                    PID_IDENTIFY_DEVICE,
-                    (uint8_t*) &g_rdm_responder.identify_on,
-                    sizeof(g_rdm_responder.identify_on));
+  g_responder.is_muted = false;
+
+  ReturnUnlessUnicast(header);
+  RDMResponder_BuildHeader(header, ACK, DISCOVERY_COMMAND_RESPONSE,
+                           PID_DISC_UN_MUTE, sizeof(uint16_t));
+  uint8_t *param_data = g_rdm_buffer + sizeof(RDMHeader);
+  param_data[0] = 0;  // set control field to 0
+  param_data[1] = 0;
+  return RDMUtil_AppendChecksum(g_rdm_buffer);
 }
 
-static void SetIdentifyDevice(const RDMHeader *header,
-                              const uint8_t *param_data,
-                              unsigned int param_data_length) {
-  uint8_t identify_on;
-  if (param_data_length != sizeof(identify_on)) {
-    SendNack(header, NR_FORMAT_ERROR);
-    return;
+int RDMResponder_GetProductDetailIds(const RDMHeader *header,
+                                     UNUSED const uint8_t *param_data) {
+  if (header->param_data_length) {
+    return RDMResponder_BuildNack(header, NR_FORMAT_ERROR);
   }
-  identify_on = param_data[0];
-  switch (identify_on) {
-    case 0:
-      g_rdm_responder.identify_on = false;
-      PLIB_PORTS_PinClear(
-        PORTS_ID_0, g_rdm_responder.identify_port,
-        g_rdm_responder.identify_bit);
-      break;
-    case 1:
-      g_rdm_responder.identify_on = true;
-      g_rdm_responder.identify_timer = CoarseTimer_GetTime();
-      PLIB_PORTS_PinSet(
-        PORTS_ID_0, g_rdm_responder.identify_port,
-        g_rdm_responder.identify_bit);
-      break;
-    default:
-      SendNack(header, NR_DATA_OUT_OF_RANGE);
-      return;
-  }
-
-  RespondIfRequired(header, ACK, SET_COMMAND_RESPONSE,
-                    PID_IDENTIFY_DEVICE, NULL, 0);
-}
 
-/*
- * @brief Reply with a string
- * @param header
- * @param param_data_length
- * @param reply_string The string to reply with
- * @param string_size The length of the string, including the terminating NULL.
- */
-static void GenericReturnString(const RDMHeader *header,
-                                unsigned int param_data_length,
-                                const char *reply_string,
-                                unsigned int string_size) {
-  if (param_data_length != 0) {
-    SendNack(header, NR_FORMAT_ERROR);
-    return;
+  ReturnUnlessUnicast(header);
+
+  const ResponderDefinition *definition = g_responder.def;
+  unsigned int size = 0;
+  unsigned int offset = sizeof(RDMHeader);
+  if (definition->product_detail_ids) {
+    unsigned int i = 0;
+    while (i < min(definition->product_detail_ids->size, MAX_PRODUCT_DETAILS)) {
```

Comment:
Okay, it just seems more fixed with a for loop, you'd question someone fiddling with i inside the loop.